contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -888,18 +923,287 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
 }
 
+/*
+ * get_immediate_base_type_of_UDT_internal()
+ * This function returns the Immediate base type for UDT.
+ * Returns InvalidOid if given type is not an UDT
+ */
+static Oid
+get_immediate_base_type_of_UDT_internal(Oid typeid)
+{
+	HeapTuple					tuple;
+	bool						isnull;
+	Datum						datum;
+	Datum                       tsql_typename;
+	Oid							base_type;
+	LOCAL_FCINFO(fcinfo, 1);
+
+	/* if common_utility_plugin_ptr is not initialised */
+	if (common_utility_plugin_ptr == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+					errmsg("Failed to find common utility plugin.")));
+
+	/* if tsql_typename is NULL it implies that inputTypId corresponds to UDT */
+	InitFunctionCallInfoData(*fcinfo, NULL, 0, InvalidOid, NULL, NULL);
+	fcinfo->args[0].value = ObjectIdGetDatum(typeid);
+	fcinfo->args[0].isnull = false;
+	tsql_typename = (*common_utility_plugin_ptr->translate_pg_type_to_tsql) (fcinfo);
+
+	/* if given type is not an UDT then return InvalidOid */
+	if (tsql_typename)
+		return InvalidOid;
+
+	/* Get immediate base type id of given type id */
+	tuple = SearchSysCache1(TYPEOID, ObjectIdGetDatum(typeid));
+	if (!HeapTupleIsValid(tuple))
+		return InvalidOid;
+
+	datum = SysCacheGetAttr(TYPEOID, tuple, Anum_pg_type_typbasetype, &isnull);
+	if (isnull)
+		return InvalidOid;
+
+	base_type = DatumGetObjectId(datum);
+	ReleaseSysCache(tuple);
+
+	return base_type;
+}
+
+Datum
+get_immediate_base_type_of_UDT(PG_FUNCTION_ARGS)
+{
+	Oid			base_type;
+	
+	base_type = get_immediate_base_type_of_UDT_internal(PG_GETARG_OID(0));
+	if (!OidIsValid(base_type))
+		PG_RETURN_NULL();
+
+	PG_RETURN_OID(base_type);
+}
+
+void
+init_special_function_list()
+{
+	Oid			type_id;
+
+	/* if common_utility_plugin_ptr is not initialised */
+	if (common_utility_plugin_ptr == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				errmsg("Failed to find common utility plugin.")));
+
+	/* mark the special function list initialised */
+	inited_tsql_special_function_list = true;
+
+	for (int special_func_idx = 0; special_func_idx < TOTAL_TSQL_SPECIAL_FUNCTION_COUNT; special_func_idx++)
+	{
+		for (int arg_idx = 0; arg_idx < tsql_special_function_list[special_func_idx].nargs; arg_idx++)
+		{
+			for (int valid_type_idx = 0; valid_type_idx < tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].len; valid_type_idx++)
+			{
+				if (!OidIsValid(tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].valid_types_oid[valid_type_idx]))
+				{
+					type_id = (*common_utility_plugin_ptr->get_tsql_datatype_oid)(tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].valid_types[valid_type_idx]);
+
+					if (OidIsValid(type_id))
+					{
+						tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].valid_types_oid[valid_type_idx] = type_id;
+					}
+					else
+					{
+						/* type id is not loaded. wait for next scan */
+						inited_tsql_special_function_list = false;
+					}
```

Comment:
Interesting! why is this handling needed?

Reply:
refer: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2495#discussion_r1684173693